File: src/main/java/com/gusrylmubarok/reddit/backend/service/impl/RefreshTokenServiceImpl.java
package com.gusrylmubarok.reddit.backend.service.impl;

import com.gusrylmubarok.reddit.backend.exceptions.BackendRedditException;
import com.gusrylmubarok.reddit.backend.model.RefreshToken;
import com.gusrylmubarok.reddit.backend.repository.RefreshTokenRepository;
import com.gusrylmubarok.reddit.backend.service.RefreshTokenService;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.UUID;

@Service
@AllArgsConstructor
@Transactional
public class RefreshTokenServiceImpl implements RefreshTokenService {

    private final RefreshTokenRepository refreshTokenRepository;

    @Override
    public RefreshToken generateRefreshToken() {
        RefreshToken refreshToken = new RefreshToken();
        refreshToken.setToken(UUID.randomUUID().toString());
        refreshToken.setCreationDate(Instant.now());

        return refreshTokenRepository.save(refreshToken);
    }

    @Override
    public void validateRefreshToken(String token) {
        refreshTokenRepository.findByToken(token)
                .orElseThrow(() -> new BackendRedditException("Invalid refresh Token"));
    }

    @Override
    public void deleteRefreshToken(String token) {
        refreshTokenRepository.deleteByToken(token);
    }
}

File: src/main/java/com/reddot/service/impl/RefreshTokenServiceImpl.kt
package com.reddot.service.impl

import com.reddot.entity.RefreshToken
import com.reddot.exception.BadRequestException
import com.reddot.repository.RefreshTokenRepository
import com.reddot.service.RefreshTokenService
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.*

@Service
class RefreshTokenServiceImpl(
    private val refreshTokenRepository: RefreshTokenRepository
) : RefreshTokenService {
    override fun generateRefreshToken(): RefreshToken {
        val token = RefreshToken(
            id = null,
            token = UUID.randomUUID().toString(),
            createDate = Instant.now()
        )
        return refreshTokenRepository.save(token)
    }

    override fun validateRefreshToken(token: String) {
        val token = refreshTokenRepository.findByToken(token)
        if (token.isEmpty) {
            throw BadRequestException("Invalid token")
        }
    }

    override fun deleteRefreshToken(token: String) {
        refreshTokenRepository.deleteByToken(token)
    }
}